### ISSUE TITLE
Platform `centos` not found (Mixlib::Install::Backend::ArtifactsNotFound)

### ISSUE THREAD

**Issue by 4-20ma** [chef-cookbooks/chef_client_updater#31]
### Cookbook version
2.0.1

### Chef-client version
13.1.31

### Platform Details
Centos 6.5

### Scenario:
Update chef-client (or downgrade chef-client); neither work.

### Steps to Reproduce:
Add `chef_client_updater::default` to Policyfile run_list; run `chef update`, `chef push staging`, chef-client will fail on node with `Mixlib::Install::Backend::ArtifactsNotFound` error.

### Expected Result:
chef-client should upgrade/downgrade as indicated in the cookbook.

### Actual Result:
Fails with `Mixlib::Install::Backend::ArtifactsNotFound` error. It appears to be related to the platform being called `centos` instead of `el`. When I load the `mixlib/install` gem in an irb session and search for chef packages, it properly returns an array of packages, but they use `el` as platform instead of `centos` or `rhel`. Should the cookbook convert the platform to a standardized string (ref. to line 76 of chef_client_updater/providers/default.rb)?

Trace below:
```
Recipe: chef_client_updater::default
  * chef_client_updater[update chef-client] action update[2017-06-17T11:31:42-05:00] WARN: NO /opt/chef.upgrade DIR TO ROLL BACK TO!

    
    ================================================================================
    Error executing action `update` on resource 'chef_client_updater[update chef-client]'
    ================================================================================
    
    Mixlib::Install::Backend::ArtifactsNotFound
    -------------------------------------------
    No artifacts found matching criteria.
      product name: chef
      channel: stable
      version: 12.19.36
      platform: centos
      platform version: 6.5
      architecture: x86_64
      compatibility mode: true
    
    Cookbook Trace:
    ---------------
    /var/chef/cache/cookbooks/chef_client_updater/providers/default.rb:106:in `update_necessary?'
    /var/chef/cache/cookbooks/chef_client_updater/providers/default.rb:221:in `block in class_from_file'
    
    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/chef_client_updater/recipes/default.rb
    
     20: chef_client_updater 'update chef-client' do
     21:   channel node['chef_client_updater']['channel']
     22:   version node['chef_client_updater']['version']
     23:   prevent_downgrade node['chef_client_updater']['prevent_downgrade']
     24:   post_install_action node['chef_client_updater']['post_install_action']
     25:   download_url_override node['chef_client_updater']['download_url_override'] if node['chef_client_updater']['download_url_override']
     26:   checksum node['chef_client_updater']['checksum'] if node['chef_client_updater']['checksum']
     27: end
    
    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/chef_client_updater/recipes/default.rb:20:in `from_file'
    
    chef_client_updater("update chef-client") do
      action [:update]
      default_guard_interpreter :default
      declared_type :chef_client_updater
      cookbook_name "chef_client_updater"
      recipe_name "default"
      channel "stable"
      version "12.19.36"
      prevent_downgrade false
      post_install_action "exec"
    end
    
    System Info:
    ------------
    chef_version=13.1.31
    platform=centos
    platform_version=6.5
    ruby=ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]
    program_name=chef-client worker: ppid=775247;start=11:31:16;
    executable=/opt/chef/bin/chef-client
    

Running handlers:
[2017-06-17T11:31:42-05:00] ERROR: Running exception handlers
Running handlers complete
[2017-06-17T11:31:42-05:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources updated in 25 seconds
[2017-06-17T11:31:43-05:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2017-06-17T11:31:43-05:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2017-06-17T11:31:43-05:00] ERROR: chef_client_updater[update chef-client] (chef_client_updater::default line 20) had an error: Mixlib::Install::Backend::ArtifactsNotFound: No artifacts found matching criteria.
  product name: chef
  channel: stable
  version: 12.19.36
  platform: centos
  platform version: 6.5
  architecture: x86_64
  compatibility mode: true

[2017-06-17T11:31:43-05:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

**Comment by 4-20ma** [comment]
### Update
Test whether changing platform string from `centos` to `el` resolves the issue. Forked cookbook and hard-coded platform to `el`, resulting in success (chef-client was properly downgraded to v12.19.36 and then subsequently upgraded to `latest` v13.1.31 by adjusting `node['chef_client_updater']['version']`).

``` diff
diff --git a/providers/default.rb b/providers/default.rb
index 7d0a2c9..2169b11 100644
--- a/providers/default.rb
+++ b/providers/default.rb
@@ -73,7 +73,7 @@ def mixlib_install
   options = {
     product_name: 'chef',
     platform_version_compatibility_mode: true,
-    platform: node['platform'],
+    platform: 'el',
     platform_version: node['platform_version'],
     architecture: node['kernel']['machine'],
     channel: new_resource.channel.to_sym,
```

**Comment by lamont-granquist** [comment]
IMO that should get fixed over in mixlib-install and we should normalize all the platforms over there with ohai platform names (which at one point i thought i tried to do back when it was all just omnitruck)

**Comment by wrightp** [comment]
mixlib install will accept any override value for platform. The correct thing to do here is remove platform, platform_version and architecture options and let mixlib install auto detect the appropriate values.

**Comment by lamont-granquist** [comment]
we had to put that here for windows.

**Comment by wrightp** [comment]
Ok I see. I'm not used to seeing this since most scenarios like this use chef-ingredient (which I understand why we're not).  May I suggest doing similar to how this is handled in chef-ingredient?

https://github.com/chef-cookbooks/chef-ingredient/blob/master/libraries/helpers.rb#L299-L310

Essentially, call `Mixlib::Install.detect_platform` and go from there.

**Comment by wrightp** [comment]
I've also see this Windows logic in omnibus: https://github.com/chef-cookbooks/omnibus/blob/master/recipes/_omnibus_toolchain.rb#L20

**Comment by 4-20ma** [comment]
@lamont-granquist I'm happy to pinch the referenced code from the chef-ingredient cookbook. Would you entertain a PR with that change? I'm not sure how this would affect the windows support.

**Comment by lamont-granquist** [comment]
yeah it looks like we should copy that logic and probably the windows logic out of the omnibus cookbook

**Comment by dguertin** [comment]
I'm still seeing the same issue with chef_client_updater 2.0.2 on Centos 7.3 nodes, trying to update from chef 12.21 to chef 13. It's the same problem we had with version 2.0.1: 

> chef_client_updater[update chef-client] (chef_client_updater::default line 20) had an error: Mixlib::Install::Backend::ArtifactsNotFound: No artifacts found matching criteria

Stacktrace follows:
`Generated at 2017-06-27 10:25:52 -0400
Mixlib::Install::Backend::ArtifactsNotFound: chef_client_updater[update chef-client] (chef_client_updater::default line 20) had an error: Mixlib::Install::Backend::ArtifactsNotFound: No artifacts found matching criteria.
  product name: chef
  channel: stable
  version: latest
  platform: centos
  platform version: 7.3.1611
  architecture: x86_64
  compatibility mode: true

/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/mixlib-install-3.3.0/lib/mixlib/install/backend/base.rb:125:in `filter_artifacts'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/mixlib-install-3.3.0/lib/mixlib/install/backend/base.rb:62:in `info'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/mixlib-install-3.3.0/lib/mixlib/install/backend.rb:29:in `info'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/mixlib-install-3.3.0/lib/mixlib/install.rb:46:in `artifact_info'
/var/chef/cache/cookbooks/chef_client_updater/providers/default.rb:111:in `update_necessary?'
/var/chef/cache/cookbooks/chef_client_updater/providers/default.rb:227:in `block in class_from_file'
(eval):2:in `block in action_update'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/provider.rb:361:in `instance_eval'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/provider.rb:361:in `compile_and_converge_action'
(eval):2:in `action_update'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/provider.rb:145:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource.rb:622:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/runner.rb:69:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/runner.rb:97:in `block (2 levels) in converge'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/runner.rb:97:in `each'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/runner.rb:97:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/runner.rb:96:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/client.rb:715:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/client.rb:710:in `catch'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/client.rb:710:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/client.rb:749:in `converge_and_save'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/client.rb:286:in `run'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:295:in `block in fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:283:in `fork'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:283:in `fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:248:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/local_mode.rb:44:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:236:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application/client.rb:464:in `sleep_then_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application/client.rb:451:in `block in interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application/client.rb:450:in `loop'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application/client.rb:450:in `interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application/client.rb:434:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:59:in `run'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/bin/chef-client:26:in `<top (required)>'
/usr/bin/chef-client:57:in `load'
/usr/bin/chef-client:57:in `<main>'

>>>> Caused by Mixlib::Install::Backend::ArtifactsNotFound: No artifacts found matching criteria.
  product name: chef
  channel: stable
  version: latest
  platform: centos
  platform version: 7.3.1611
  architecture: x86_64
  compatibility mode: true

/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/mixlib-install-3.3.0/lib/mixlib/install/backend/base.rb:125:in `filter_artifacts'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/mixlib-install-3.3.0/lib/mixlib/install/backend/base.rb:62:in `info'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/mixlib-install-3.3.0/lib/mixlib/install/backend.rb:29:in `info'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/mixlib-install-3.3.0/lib/mixlib/install.rb:46:in `artifact_info'
/var/chef/cache/cookbooks/chef_client_updater/providers/default.rb:111:in `update_necessary?'
/var/chef/cache/cookbooks/chef_client_updater/providers/default.rb:227:in `block in class_from_file'
(eval):2:in `block in action_update'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/provider.rb:361:in `instance_eval'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/provider.rb:361:in `compile_and_converge_action'
(eval):2:in `action_update'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/provider.rb:145:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource.rb:622:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/runner.rb:69:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/runner.rb:97:in `block (2 levels) in converge'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/runner.rb:97:in `each'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/runner.rb:97:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/runner.rb:96:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/client.rb:715:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/client.rb:710:in `catch'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/client.rb:710:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/client.rb:749:in `converge_and_save'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/client.rb:286:in `run'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:295:in `block in fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:283:in `fork'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:283:in `fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:248:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/local_mode.rb:44:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:236:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application/client.rb:464:in `sleep_then_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application/client.rb:451:in `block in interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application/client.rb:450:in `loop'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application/client.rb:450:in `interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application/client.rb:434:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/lib/chef/application.rb:59:in `run'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.21.1/bin/chef-client:26:in `<top (required)>'
/usr/bin/chef-client:57:in `load'
/usr/bin/chef-client:57:in `<main>'                                                                                                                                                                                                                                    
`

**Comment by tas50** [comment]
This is fixed with the 2.0.3 release
